// ember-electron/lib/commands/electron-test/task.js
'use strict'

const path = require('path')
const getElectronApp = require('../../helpers/find-electron').getElectronApp
const safePath = require('../../helpers/safe-path')

module.exports = {
  testemOptions: function (options) {
    /* jscs:disable requireCamelCaseOrUpperCaseIdentifiers */
    return {
      host: options.host,
      port: options.port,
      cwd: options.outputPath,
      reporter: options.reporter,
      middleware: this.addonMiddlewares(),
      launch_in_dev: ['Electron'],
      launch_in_ci: ['Electron (CI)'],
      launchers: {
        'Electron (CI)': {
          command: this.electronCommand(options, 'ci'),
          protocol: 'tap'
        },
        Electron: {
          command: this.electronCommand(options, 'dev'),
          protocol: 'browser'
        }
      }
    }
  /* jscs:enable requireCamelCaseOrUpperCaseIdentifiers */
  },

  electronCommand: function (options, type) {
    const dirname = __dirname || path.resolve(path.dirname())
    const command = 'node'
    const commandArgs = safePath(path.join(dirname, `./runner-${type}.js`))
    const commandFlags = [
      '--electron-path ' + safePath(getElectronApp(this.project)),
      '--tests-path ' + safePath(path.join(options.outputPath, 'tests'))
    ]

    return [command, commandArgs].concat(commandFlags).join(' ')
  }
}
