// ember-electron/lib/helpers/find-electron.js
'use strict'

const path = require('path')
const os = require('os')
const dirExists = require('./directory-exists')

/**
 * Takes a path to an Electron prebuilt node_modules folder
 * (electron-prebuilt or electron) and returns the path to
 * the executable, depending on the current platform
 *
 * @param {string} electronLocalPath
 * @returns {string} Path to executable
 */
function getLocalElectron (electronLocalPath) {
  const platform = os.platform()

  if (platform === 'darwin') {
    return path.join(electronLocalPath, 'Electron.app', 'Contents', 'MacOS', 'Electron')
  }

  if (platform === 'win32') {
    return path.join(electronLocalPath, 'electron.exe')
  }

  return path.join(electronLocalPath, 'electron')
}

/**
 * Checks if electron or electron-prebuilt is installed in
 * node_modules and returns the path to the binary
 *
 * @param {Object} project
 * @returns {string} Path to Electron executable
 */
function getElectronApp (project) {
  const localElectron = getElectronPackagePath(project)
  let electron

  if (localElectron) {
    electron = getLocalElectron(localElectron)
  } else if (process.env.ELECTRON_PATH) {
    electron = process.env.ELECTRON_PATH
  } else {
    electron = 'Electron'
  }

  return electron
}

/**
 * Checks if electron or electron-prebuilt is installed in
 * node_modules and returns the path to either package
 *
 * @param {Object} project
 * @returns {string} Path to local electron or electron-prebuilt
 */
function getElectronPackagePath (project) {
  const localElectrons = {
    electron: {
      path: path.resolve(project.root, 'node_modules/electron-prebuilt/dist/'),
      exists: null
    },
    prebuilt: {
      path: path.resolve(project.root, 'node_modules/electron/dist/'),
      exists: null
    }
  }

  if (dirExists(localElectrons.electron.path)) return localElectrons.electron.path
  if (dirExists(localElectrons.prebuilt.path)) return localElectrons.prebuilt.path
  return null
}

module.exports = {
  getElectronPackagePath: getElectronPackagePath,
  getLocalElectron: getLocalElectron,
  getElectronApp: getElectronApp
}
